### ISSUE TITLE
Minikube: error converting pod to perceptor pod: unable to parse kubernetes imageID string

### ISSUE THREAD

**Issue by jlin963** [blackducksoftware/perceivers#44]
I'm getting the following error in the pod-perceiver logs when running on Minikube (installed through protoform):

> E0327 14:02:13.122262       1 pod_controller.go:171] kube-system/kubernetes-dashboard-77d8b98585-9kndw failed with : error converting pod to perceptor pod: unable to parse kubernetes imageID string docker://sha256:e94d2f21bc0c297cb74c1dfdd23e2eace013f532c60726601af67984d97f718a from pod kube-system/kubernetes-dashboard-77d8b98585-9kndw: unable to match digestRegexp regex &lt;@sha256:([a-zA-Z0-9]+)$&gt; to input &lt;docker://sha256:e94d2f21bc0c297cb74c1dfdd23e2eace013f532c60726601af67984d97f718a&gt;

Setup:
-Minikube v1.6
-# of concurrent scans: 1

[logs-from-perceptor-in-perceptor-xpwsk.txt](https://github.com/blackducksoftware/perceivers/files/1852371/logs-from-perceptor-in-perceptor-xpwsk.txt)
[logs-from-perceptor-scanner-in-perceptor-scanner-ftfrz.txt](https://github.com/blackducksoftware/perceivers/files/1852372/logs-from-perceptor-scanner-in-perceptor-scanner-ftfrz.txt)
[logs-from-pod-perceiver-in-pod-perceiver-gcrg4.txt](https://github.com/blackducksoftware/perceivers/files/1852373/logs-from-pod-perceiver-in-pod-perceiver-gcrg4.txt)

**Comment by rrati** [comment]
Is that image getting annotated at some point?  Do you have a reproduction case?

**Comment by jlin963** [comment]
Not sure about the annotations; still getting familiar with Kube vs OpenShift (currently looking into that). From the logs, it looks like all the images are throwing that error (I just copy-pasted a specific instance of it). I can give steps to reproduce though I think...I went through these steps twice and saw it both times. 

Assuming you have Minikube, some VM (I used VirtualBox), and kubectl installed already:

1. Run: minikube start --vm-driver=virtualbox
2. In perceptor-protoform/install/kube1.6, run: ./install.sh -u 300m -m 300Mi -i -d -M
3. Fill in the Hub config info. I used hub-piv01.blackducksoftware.com, 443, sysadmin, and 1 concurrent scan.
4. Wait for a few minutes.

Personally I ran minikube dashboard to access the UI and looked at the logs that way, but you could just look at them through the command line.

**Comment by jlin963** [comment]
Still not sure how to look at image annotations in Kube, but here are the pod annotations: 
[pod_annotations.txt](https://github.com/blackducksoftware/perceivers/files/1852487/pod_annotations.txt)

It's been stuck on the same image for 30+ minutes though. I keep looking at the metrics and the numbers haven't updated for that time.

**Comment by jlin963** [comment]
Ah okay, so when you launch Minikube it creates a namespace called kube-system with 4 pods (kube-addon-manager-minikube, kube-dns-54cccfbdf8-t6bxt, and kubernetes-dashboard-77d8b98585-9kndw, storage-provisioner). Those are the ones whose containers are throwing that E0327 error in the pod-perceiver logs. Maybe b/c they're not in the bds-perceptor namespace? They are being picked up for scanning though (I can see storage-provisioner in my Hub scan results and that's under the kube-system namespace).

It looks like they do get annotated though (see image).

![screen shot 2018-03-27 at 11 53 17 am](https://user-images.githubusercontent.com/31659508/37979019-7f28c30a-31b5-11e8-95e2-6ac2fb99aaa6.png)

**Comment by jlin963** [comment]
Also seeing this with a pod I created and added to my cluster. 

time="2018-04-02T18:42:58Z" level=error msg="unable to parse kubernetes imageID string  from pod default/nginx-mysql: unable to match digestRegexp regex &lt;@sha256:([a-zA-Z0-9]+)$&gt; to input &lt;&gt;"

same error, just a different pod name.

**Comment by mattfenwick** [comment]
I looked into this a bit more, here are my preliminary findings:

it seems that there are some `containerStatuses` which have an imageID which is not pullable by docker, for example:

```
                "containerStatuses": [
                    {
                        "containerID": "docker://763cacceb1b531f0b63c0b8e47ded82f07f31a8e5808f0d026c382e61c71bfa7",
                        "image": "gcr.io/google-containers/kube-addon-manager:v6.5",
                        "imageID": "docker://sha256:d166ffa9201aa156eb76d3a221c3fdab07bb1a0b6407548b1b1f03dc111c0e39",
                        "lastState": {
                            "terminated": {
                                "containerID": "docker://535a9f4687563c604304656ef18ae3f5aaf8b1f09817e55f9cb74f307a11dbdc",
                                "exitCode": 137,
                                "finishedAt": "2018-03-28T20:42:30Z",
                                "reason": "Error",
                                "startedAt": "2018-03-21T15:39:42Z"
                            }
                        },
                        "name": "kube-addon-manager",
                        "ready": true,
                        "restartCount": 1,
                        "state": {
                            "running": {
                                "startedAt": "2018-04-02T19:37:49Z"
                            }
                        }
                    }
                ]
```

for other containerStatus objects, the `ImageID` is something like `"docker-pullable://gcr.io/gke-verification/blackducksoftware/echoer@sha256:80c27c87683c6e42c0dfd8999264bf8cbffb41d8562637f4693c1194340cab90"` where the `80c27c87683c6e42c0dfd8999264bf8cbffb41d8562637f4693c1194340cab90` is a valid docker image sha.  

In the example container status above, `d166ffa9201aa156eb76d3a221c3fdab07bb1a0b6407548b1b1f03dc111c0e39` might not be a valid docker image sha:

```
$ docker pull gcr.io/google-containers/kube-addon-manager@sha256:d166ffa9201aa156eb76d3a221c3fdab07bb1a0b6407548b1b1f03dc111c0e39
Error response from daemon: manifest for gcr.io/google-containers/kube-addon-manager@sha256:d166ffa9201aa156eb76d3a221c3fdab07bb1a0b6407548b1b1f03dc111c0e39 not found
```

I'm not sure what the action item is here ... how do we get the image shas if they're not provided in the container status?

**Comment by mattfenwick** [comment]
It seems that kubernetes is specifying the image using the `Id` field from the docker image:

```
$ docker inspect k8s.gcr.io/k8s-dns-sidecar-amd64:1.14.5
[
    {
        "Id": "sha256:fed89e8b4248a788655d528d96fe644aff012879c782784cd486ff6894ef89f6",
        "RepoTags": [
            "k8s.gcr.io/k8s-dns-sidecar-amd64:1.14.5"
        ],
        "RepoDigests": [
            "k8s.gcr.io/k8s-dns-sidecar-amd64@sha256:9aab42bf6a2a068b797fe7d91a5d8d915b10dbbc3d6f2b10492848debfba6044"
        ],
        "Parent": "",
        "Comment": "",
        "Created": "2017-09-27T04:45:31.552746448Z",
        "Container": "5dcf93baab35d8521239adaf7f30001c3058ae33ffaed6581d1cb5f266a62654",
        "ContainerConfig": {
            "Hostname": "f0ee9a8d89fb",
            "Domainname": "",
            "User": "nobody:nobody",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "/bin/sh",
                "-c",
                "#(nop) ",
                "ENTRYPOINT [\"/sidecar\"]"
            ],
            "ArgsEscaped": true,
            "Image": "sha256:bdc7dc8ba08c2ac8c8e03550d8ebf3297a669a3f03e36c377b9515f08c1b4ef4",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": [
                "/sidecar"
            ],
            "OnBuild": [],
            "Labels": {}
        },
        "DockerVersion": "1.12.6",
        "Author": "Bowei Du <bowei@google.com>",
        "Config": {
            "Hostname": "f0ee9a8d89fb",
            "Domainname": "",
            "User": "nobody:nobody",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": null,
            "ArgsEscaped": true,
            "Image": "sha256:bdc7dc8ba08c2ac8c8e03550d8ebf3297a669a3f03e36c377b9515f08c1b4ef4",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": [
                "/sidecar"
            ],
            "OnBuild": [],
            "Labels": {}
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 41819177,
        "VirtualSize": 41819177,
        "GraphDriver": {
            "Data": null,
            "Name": "aufs"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:5bef08742407efd622d243692b79ba0055383bbce12900324f75e56f589aedb0",
                "sha256:acfc450a47fa9d10a45cb63e1437c7fd0d6cda51810e0a485e29c59baf94e0a0"
            ]
        },
        "Metadata": {
            "LastTagTime": "0001-01-01T00:00:00Z"
        }
    }
]
```

**Comment by rrati** [comment]
@jayunit100 description would make sense.  Static pods as they are called produce mirror pods in the apiserver.  If that is the case, and it makes sense, then we probably need to figure out a way to properly handle this or silently ignore them.

**Comment by rrati** [comment]
Maybe we need to adjust the regex because the docker://sha256@ is similar to an image in openshift.

**Comment by msenmurugan** [comment]
Two parts here ; 
1) https://github.com/blackducksoftware/perceivers/pull/48/
2) Afterwards, for UX, lets roll into a `map[string]int` histogram and print the *Whole* histogram out only if we;'re seeing a new *key*. i.e. 
```
cant-pull-this:sha://sodjf02e98j02e9uj0239jd2039jd, 
/f8i0283ur0d9j2039dj203ejd
/f8i0283ur0d9j2039dj203ejd
```
In this case, we'd have a histogram of 
```
cant-pull-this:sha://sodjf02e98j02e9uj0239jd2039jd,  1
/f8i0283ur0d9j2039dj203ejd, 2
```
And logs would have exactly 2 printouts,.

**Comment by jayunit100** [comment]
I dont think this is a 1x blocker necessarily.  is it?

**Comment by rrati** [comment]
fixed via #52
